### PR TITLE
MIGRATION WILL CAUSE DOWN TIME: Add a sharing_allowed column to the user table

### PR DESCRIPTION
**Pull request by caleybrock**
Internal spec: https://docs.google.com/document/d/1ISuKetN3uEe6XDPM4VIZRM239euKotnd3p7Eoi2B_ls/edit#

Adding a sharing_allow column to the user table. This flag will determine wether uses are allowed to share their projects. Teachers will be able to control this. It should always be true unless teachers set it to false. 

This is a column instead of a property because Poorva expected to run a query on it about once a month.

### CODE REVIEW

**Comment by caleybrock** [comment]
Should I even be merging this into staging? Does it have to go through test first or does this get treated like a hotfix and then backported?

**Comment by ashercodeorg** [comment]
This should get merged to `staging` exactly when it is ready to be moved through `test` to `production`. Maybe we pick an evening this week to do it together? Note that the change to `schema_cache.dump` can be grabbed from `staging` when DTT'ing (generated by the build system) or can be added to this PR.

File: dashboard/db/migrate/20170814203015_add_share_flag_to_user.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class AddShareFlagToUser < ActiveRecord::Migration[5.0]
+  def change
+    add_column :users, :sharing_allowed, :boolean, null: false, default: true
```
Comment (ashercodeorg) [inline dashboard/db/migrate/20170814203015_add_share_flag_to_user.rb:3]: nit: Does this column belong at the end of the `User` model? Alternately after some other particular column (e.g., via [this syntax](https://github.com/code-dot-org/code-dot-org/blob/staging/dashboard/db/migrate/20170619170358_add_purged_at_to_users.rb#L3))?

**Comment by caleybrock** [comment]
Sure, let's tentatively say Wednesday and I'll try to have a green DTT ready to go by 5pm? or earlier? 
cc @jeremydstone who will be DOTD that day.

File: dashboard/db/migrate/20170814203015_add_share_flag_to_user.rb (hunk 1 above)
Comment (ashercodeorg) [reply dashboard/db/migrate/20170814203015_add_share_flag_to_user.rb:3]: How is this going to be used? Do we expect to ever search based on `sharing_allowed`? Does it make sense to add an index on it?

Comment (ashercodeorg) [reply dashboard/db/migrate/20170814203015_add_share_flag_to_user.rb:3]: nit: Mind adding a comment as to the meaning of this column, either via ruby comment or via DB comment or both?

**Comment by aoby** [comment]
I like the idea of a UserOptions table. I think in the past we've not done it because that can be somewhat covered by the existing `User.properties`. The obvious downside here is that we want to query (periodically, ~monthly) users on this field. Maybe this could be the catalyst for starting a UserOptions table?

File: dashboard/db/migrate/20170814203015_add_share_flag_to_user.rb (hunk 1 above)
Comment (aoby) [reply dashboard/db/migrate/20170814203015_add_share_flag_to_user.rb:3]: For the comment, we should add a [Rails 5 migration comment](http://blog.bigbinary.com/2016/06/21/rails-5-supports-adding-comments-migrations.html), which will persist in the DB itself and in the schema.rb

**Comment by caleybrock** [comment]
I've documented brainstorming notes, as well as a technical plan approved by Jeremy/Asher/Poorva here: https://docs.google.com/document/d/1zKssQHvRM5T53BDs_ZpYhEZek7Eftiv66-9PHldkBTc/edit#

Closing this PR and making a new one to add it to user properties.

**Pull request closed without merging**